qiskit/circuit/measure.py

```diff
@@ -14,15 +14,21 @@
 Quantum measurement in the computational basis.
 """
 
+from __future__ import annotations
+
 import warnings
 
 from qiskit.circuit.instruction import Instruction
 from qiskit.circuit.exceptions import CircuitError
 
 
+# Measure class kept for backwards compatibility, and repurposed
+# as a common parent class for all measurement instructions.
 class Measure(Instruction):
```

Comment:
It's _possible_ that we could accept an inheritance relationship here, but I don't think it's for certain the correct way of doing things, over (for example) a set of flags defined `Instruction` or `Operation`. This particular inheritance is definitely not safe, though - `Measure` is, and needs to remain, a measure in the Z basis, so `MeasureX` isn't an instance of it.  There's also the false abstraction here that the base class is only abstract as far as a single-qubit Pauli-basis measurement, which isn't the most general type.

Aside from those abstraction/false-hierarchy issues, another thing to consider with inheritance is that we would need to work out how external / further subclasses of `BaseMeasure` (or whatever) would be safely serialised/deserialised by QPY - this PR adds an inheritance structure to instructions that is deliberately avoided in our data model elsewhere. This has some wide-ranging implications that I'm too "on holiday" to think through properly right now (sorry).

Reply:
I see 🤔 let's pick this up again once you come back then! Thanks for the detailed explanation